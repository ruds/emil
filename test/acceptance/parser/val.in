val foo = bar;
val foo = bar.baz;
val foo as (num, str) = (1, "bar");
val (num, _) = (1, "bar");
val _ = foo;
val (++) = (+);
val (prefix +) = fn x => x; (* TODO -- make non-error when fn implemented *)
val foo.bar.baz = t; (* parses OK - not a very useful assignment *)
val { foo, bar as (baz, quux), waldo = 2, ... } = xyzzy;
val () = foo;
val (foo, _, (++), (=), (*), bar) = baz;
val [] = foo;
val [foo as (_, bar), baz, {quux, waldo = (xyzzy, plugh)}, _] = grault;
val Foo bar = baz;
val (::) (foo, bar) = baz;
val (prefix ~) foo = bar;
(* add parenthesized pattern, operator precedence tests *)
val (_, 2.3) = foo; (* error - fp literals not allowed *)
val;;
val foo;
val foo =;;
