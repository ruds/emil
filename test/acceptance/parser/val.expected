0001 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((IdentifierPattern
                 :qualifiers
                 ()
                 :identifier
                 "666F6F"
                 :is_op
                 false
                 :is_prefix_op
                 false)
               . (IdentifierExpr
                     :qualifiers
                     ()
                     :identifier
                     "626172"
                     :is_op
                     false
                     :is_prefix_op
                     false)))))
0002 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((IdentifierPattern
                 :qualifiers
                 ()
                 :identifier
                 "666F6F"
                 :is_op
                 false
                 :is_prefix_op
                 false)
               . (IdentifierExpr
                     :qualifiers
                     ("626172")
                     :identifier
                     "62617A"
                     :is_op
                     false
                     :is_prefix_op
                     false)))))
0003 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((LayeredPattern
                 :identifier
                 "666F6F"
                 :pattern
                 (TuplePattern
                     :patterns
                     (
                         (IdentifierPattern
                             :qualifiers
                             ()
                             :identifier
                             "6E756D"
                             :is_op
                             false
                             :is_prefix_op
                             false)
                         (IdentifierPattern
                             :qualifiers
                             ()
                             :identifier
                             "737472"
                             :is_op
                             false
                             :is_prefix_op
                             false)
                     )))
               . (TupleExpr
                     :exprs
                     (
                         (BigintLiteralExpr
                             :val
                             1)
                         (StringLiteralExpr
                             :val
                             "626172")
                     ))))))
0004 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((TuplePattern
                 :patterns
                 (
                     (IdentifierPattern
                         :qualifiers
                         ()
                         :identifier
                         "6E756D"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     (WildcardPattern)
                 ))
               . (TupleExpr
                     :exprs
                     (
                         (BigintLiteralExpr
                             :val
                             1)
                         (StringLiteralExpr
                             :val
                             "626172")
                     ))))))
0005 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((WildcardPattern)
               . (IdentifierExpr
                     :qualifiers
                     ()
                     :identifier
                     "666F6F"
                     :is_op
                     false
                     :is_prefix_op
                     false)))))
0006 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((IdentifierPattern
                 :qualifiers
                 ()
                 :identifier
                 "2B2B"
                 :is_op
                 true
                 :is_prefix_op
                 false)
               . (IdentifierExpr
                     :qualifiers
                     ()
                     :identifier
                     "2B"
                     :is_op
                     true
                     :is_prefix_op
                     false)))))
0007 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((IdentifierPattern
                 :qualifiers
                 ()
                 :identifier
                 "2B"
                 :is_op
                 true
                 :is_prefix_op
                 true)
               . (FnExpr
                     :cases
                     (((IdentifierPattern
                         :qualifiers
                         ()
                         :identifier
                         "78"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                       . (IdentifierExpr
                             :qualifiers
                             ()
                             :identifier
                             "78"
                             :is_op
                             false
                             :is_prefix_op
                             false))))))))
0008 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((IdentifierPattern
                 :qualifiers
                 (
                     "666F6F"
                     "626172"
                 )
                 :identifier
                 "62617A"
                 :is_op
                 false
                 :is_prefix_op
                 false)
               . (IdentifierExpr
                     :qualifiers
                     ()
                     :identifier
                     "74"
                     :is_op
                     false
                     :is_prefix_op
                     false)))))
0009 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((PolyRecordPattern
                 :rows
                 (
                     (RecRowPattern
                         :label
                         "666F6F"
                         :pattern
                         (IdentifierPattern
                             :qualifiers
                             ()
                             :identifier
                             "666F6F"
                             :is_op
                             false
                             :is_prefix_op
                             false))
                     (RecRowPattern
                         :label
                         "626172"
                         :pattern
                         (LayeredPattern
                             :identifier
                             "626172"
                             :pattern
                             (TuplePattern
                                 :patterns
                                 (
                                     (IdentifierPattern
                                         :qualifiers
                                         ()
                                         :identifier
                                         "62617A"
                                         :is_op
                                         false
                                         :is_prefix_op
                                         false)
                                     (IdentifierPattern
                                         :qualifiers
                                         ()
                                         :identifier
                                         "71757578"
                                         :is_op
                                         false
                                         :is_prefix_op
                                         false)
                                 ))))
                     (RecRowPattern
                         :label
                         "77616C646F"
                         :pattern
                         (LiteralPattern
                             :val
                             (BigintLiteralExpr
                                 :val
                                 2)))
                 ))
               . (IdentifierExpr
                     :qualifiers
                     ()
                     :identifier
                     "78797A7A79"
                     :is_op
                     false
                     :is_prefix_op
                     false)))))
0010 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((TuplePattern
                 :patterns
                 ())
               . (IdentifierExpr
                     :qualifiers
                     ()
                     :identifier
                     "666F6F"
                     :is_op
                     false
                     :is_prefix_op
                     false)))))
0011 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((TuplePattern
                 :patterns
                 (
                     (IdentifierPattern
                         :qualifiers
                         ()
                         :identifier
                         "666F6F"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     (WildcardPattern)
                     (IdentifierPattern
                         :qualifiers
                         ()
                         :identifier
                         "2B2B"
                         :is_op
                         true
                         :is_prefix_op
                         false)
                     (IdentifierPattern
                         :qualifiers
                         ()
                         :identifier
                         "3D"
                         :is_op
                         true
                         :is_prefix_op
                         false)
                     (IdentifierPattern
                         :qualifiers
                         ()
                         :identifier
                         "2A"
                         :is_op
                         true
                         :is_prefix_op
                         false)
                     (IdentifierPattern
                         :qualifiers
                         ()
                         :identifier
                         "626172"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                 ))
               . (IdentifierExpr
                     :qualifiers
                     ()
                     :identifier
                     "62617A"
                     :is_op
                     false
                     :is_prefix_op
                     false)))))
0012 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((ListPattern
                 :patterns
                 ())
               . (IdentifierExpr
                     :qualifiers
                     ()
                     :identifier
                     "666F6F"
                     :is_op
                     false
                     :is_prefix_op
                     false)))))
0013 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((ListPattern
                 :patterns
                 (
                     (LayeredPattern
                         :identifier
                         "666F6F"
                         :pattern
                         (TuplePattern
                             :patterns
                             (
                                 (WildcardPattern)
                                 (IdentifierPattern
                                     :qualifiers
                                     ()
                                     :identifier
                                     "626172"
                                     :is_op
                                     false
                                     :is_prefix_op
                                     false)
                             )))
                     (IdentifierPattern
                         :qualifiers
                         ()
                         :identifier
                         "62617A"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     (RecordPattern
                         :rows
                         (
                             (RecRowPattern
                                 :label
                                 "71757578"
                                 :pattern
                                 (IdentifierPattern
                                     :qualifiers
                                     ()
                                     :identifier
                                     "71757578"
                                     :is_op
                                     false
                                     :is_prefix_op
                                     false))
                             (RecRowPattern
                                 :label
                                 "77616C646F"
                                 :pattern
                                 (TuplePattern
                                     :patterns
                                     (
                                         (IdentifierPattern
                                             :qualifiers
                                             ()
                                             :identifier
                                             "78797A7A79"
                                             :is_op
                                             false
                                             :is_prefix_op
                                             false)
                                         (IdentifierPattern
                                             :qualifiers
                                             ()
                                             :identifier
                                             "706C756768"
                                             :is_op
                                             false
                                             :is_prefix_op
                                             false)
                                     )))
                         ))
                     (WildcardPattern)
                 ))
               . (IdentifierExpr
                     :qualifiers
                     ()
                     :identifier
                     "677261756C74"
                     :is_op
                     false
                     :is_prefix_op
                     false)))))
0014 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((DatatypePattern
                 :constructor
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "466F6F"
                     :is_op
                     false
                     :is_prefix_op
                     false)
                 :arg
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "626172"
                     :is_op
                     false
                     :is_prefix_op
                     false))
               . (IdentifierExpr
                     :qualifiers
                     ()
                     :identifier
                     "62617A"
                     :is_op
                     false
                     :is_prefix_op
                     false)))))
0015 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((DatatypePattern
                 :constructor
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "3A3A"
                     :is_op
                     true
                     :is_prefix_op
                     false)
                 :arg
                 (TuplePattern
                     :patterns
                     (
                         (IdentifierPattern
                             :qualifiers
                             ()
                             :identifier
                             "666F6F"
                             :is_op
                             false
                             :is_prefix_op
                             false)
                         (IdentifierPattern
                             :qualifiers
                             ()
                             :identifier
                             "626172"
                             :is_op
                             false
                             :is_prefix_op
                             false)
                     )))
               . (IdentifierExpr
                     :qualifiers
                     ()
                     :identifier
                     "62617A"
                     :is_op
                     false
                     :is_prefix_op
                     false)))))
0016 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             (((DatatypePattern
                 :constructor
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "7E"
                     :is_op
                     true
                     :is_prefix_op
                     true)
                 :arg
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "666F6F"
                     :is_op
                     false
                     :is_prefix_op
                     false))
               . (IdentifierExpr
                     :qualifiers
                     ()
                     :identifier
                     "626172"
                     :is_op
                     false
                     :is_prefix_op
                     false)))))
0018 ERROR
0019 ERROR
0020 ERROR
0021 ERROR
0022 ERROR
0023 (EndOfFileTopDecl)
